### PR TITLE
Adds the Ringbuffer to the old client.

### PR DESCRIPTION
**Pull request by pveentjer**
This is a backport of the Ringbuffer for the old client added to the master branch.

### CODE REVIEW

**Comment by asimarslan** [comment]
:+1:

File: hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientRingbufferProxy.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        Long result = invoke(request);
+        return result;
+    }
+
+    @Override
+    public long add(E item) {
+        checkNotNull(item, "item can't be null");
+
+        AddRequest request = new AddRequest(getName(), toData(item));
+        Long result = invoke(request);
```
Comment (sancar) [inline hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientRingbufferProxy.java]: Should these requests to be send to node that owning a specific partition ? 
If yes, these seem to be send to random node. It is similar problem in #5861. 
see https://github.com/hazelcast/hazelcast/pull/5861/files#r36734618

Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientRingbufferProxy.java]: Thanks I'll have a look it and change the partition-id passing.

File: hazelcast-client/src/test/java/com/hazelcast/client/ringbuffer/RingbufferTest.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    static HazelcastInstance server;
+    private Ringbuffer<String> clientRingbuffer;
+    private Ringbuffer<String> serverRingbuffer;
+
+    @BeforeClass
+    public static void init() {
+        Config config = new Config();
+        config.addRingBufferConfig(new RingbufferConfig("rb*").setCapacity(CAPACITY));
+
+        server = Hazelcast.newHazelcastInstance(config);
```
Comment (sancar) [inline hazelcast-client/src/test/java/com/hazelcast/client/ringbuffer/RingbufferTest.java]: Mocked network for the tests ?

File: hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/client/RingbufferRequest.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        final String partitionKey = StringPartitioningStrategy.getPartitionKey(name);
+        return getClientEngine().getPartitionService().getPartitionId(partitionKey);
+    }
+
+    @Override
+    public String getServiceName() {
+        return RingbufferService.SERVICE_NAME;
+    }
+
+    //todo
```
Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/client/RingbufferRequest.java]: todo  ?!?!

**Comment by sancar** [comment]
invoke(request, getPartitionId()); is calling wrong method right now. 
I believe you meant to add following to ClientProxy but you did not yet
```
 protected <T> T invoke(ClientRequest req, int partitionId) {
        try {
            final Future future = new ClientInvocation(getClient(), req, partitionId).invoke();
            Object result = future.get();
            return toObject(result);
        } catch (Exception e) {
            throw ExceptionUtil.rethrow(e);
        }
    }
```

File: hazelcast-client/src/test/java/com/hazelcast/client/ringbuffer/RingbufferTest.java (hunk 2 above)
Comment (pveentjer) [inline hazelcast-client/src/test/java/com/hazelcast/client/ringbuffer/RingbufferTest.java]: Sure.. will change.

**Comment by pveentjer** [comment]
I'll fix that.

**Comment by pveentjer** [comment]
I have resolved the partition-id issue + test is using mock network.

**Comment by sancar** [comment]
Again, I am Ok with the pr, as long as we do not skip the security related parts before release.

**Pull request merged**